File: backend/Services/CategoryService.cs
using backend.Data;
using backend.DTOs;
using backend.Models;
using Microsoft.EntityFrameworkCore;

namespace backend.Services;

public class CategoryService
{
    private readonly ApplicationDbContext _context;

    public CategoryService(ApplicationDbContext context)
    {
        _context = context;
    }

    // GET ALL CATEGORIES (CLEAN DTO RESPONSE)
    public async Task<List<CategoryDto>> GetAll()
    {
        return await _context.Categories
            .AsNoTracking()
            .Select(c => new CategoryDto
            {
                Id = c.Id,
                Name = c.Name
            })
            .ToListAsync();
    }

    // SEARCH CATEGORIES BY NAME
    public async Task<List<CategoryDto>> SearchByName(string name)
    {
        return await _context.Categories
            .Where(c => c.Name.Contains(name))
            .AsNoTracking()
            .Select(c => new CategoryDto
            {
                Id = c.Id,
                Name = c.Name
            })
            .ToListAsync();
    }

    // CREATE NEW CATEGORY
    public async Task<CategoryDto> Create(CreateCategoryDto dto)
    {
        var category = new Category { Name = dto.Name };
        _context.Categories.Add(category);
        await _context.SaveChangesAsync();

        return new CategoryDto { Id = category.Id, Name = category.Name };
    }

    // UPDATE CATEGORY
    public async Task<CategoryDto?> Update(int id, UpdateCategoryDto dto)
    {
        var category = await _context.Categories.FindAsync(id);
        if (category == null)
            return null;

        category.Name = dto.Name;
        _context.Categories.Update(category);
        await _context.SaveChangesAsync();

        return new CategoryDto { Id = category.Id, Name = category.Name };
    }

    // DELETE CATEGORY
    public async Task<bool> Delete(int id)
    {
        var category = await _context.Categories.FindAsync(id);
        if (category == null)
            return false;

        _context.Categories.Remove(category);
        await _context.SaveChangesAsync();
        return true;
    }
}

File: frontend/src/services/categoryService.ts
import { api } from './api';
import type { Category, CreateCategoryDto, UpdateCategoryDto } from '../interfaces/Category';

export const categoryService = {
  getAll: async (): Promise<Category[]> => {
    return api.get('/category');
  },

  searchByName: async (name: string): Promise<Category[]> => {
    return api.get(`/category/search?name=${encodeURIComponent(name)}`);
  },

  create: async (data: CreateCategoryDto): Promise<Category> => {
    return api.post('/category', data);
  },

  update: async (id: number, data: UpdateCategoryDto): Promise<Category> => {
    return api.put(`/category/${id}`, data);
  },

  delete: async (id: number): Promise<void> => {
    return api.delete(`/category/${id}`);
  },
};
